src/mlpack/methods/ann/layer/concat_layer.hpp

```diff
@@ -0,0 +1,183 @@
+/**
+ * @file concat_layer.hpp
+ * @author Nilay Jain
+ * @author Marcus Edel
+ *
+ * Definition of the ConcatLayer class.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_CONCAT_LAYER_HPP
+#define MLPACK_METHODS_ANN_LAYER_CONCAT_LAYER_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/activation_functions/rectifier_function.hpp>
+#include <mlpack/methods/ann/layer/layer_traits.hpp>
+#include <mlpack/methods/ann/layer/one_hot_layer.hpp>
+#include <mlpack/methods/ann/layer/conv_layer.hpp>
+#include <mlpack/methods/ann/layer/pooling_layer.hpp>
+#include <mlpack/methods/ann/layer/softmax_layer.hpp>
+#include <mlpack/methods/ann/layer/bias_layer.hpp>
+#include <mlpack/methods/ann/layer/linear_layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+/**
+ * @tparam JoinLayers Contains all layer modules that need be concatenated.
+ */
+template <typename JoinLayers,
+          typename InputDataType = arma::cube,
+          typename OutputDataType = arma::cube>
+class ConcatLayer
+{
+ public:
+  ConcatLayer(const size_t numLayers,
+              JoinLayers &&layers) :
+  numLayers(numLayers),
+  layers(std::forward<JoinLayers>(layers))
```

Comment:
Your solution tell the compiler "JoinLayers must be rvalue", that means

```
SomeLayer layers;
ConcatLayer<SomeLayer> ConcatLayer(layers);
```
Do not work, because layers is an lvalue in this case, but the constructor only accept rvalue.

If you want to perform perfect forwarding by std::forward(which means, if it is rvalue, cast it to rvalue, if it is lvalue, cast it lvalue), you need to write it as

```
template<typename JoinLayersType>
ConcatLayer(const size_t numLayers,
                    JoinLayersType &&layers) ;
```
In this case, the compiler will treat JoinLayers&& as "[universal reference](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers)", so perfect forwarding work under this case.

You may ask, but I need the JoinLayersType == JoinLayers, use can use static_assert to add this restriction work

```
using RawJLT = typename std::decay<JoinLayersType >::type;
static_assert(std::is_same<RawJLT , JoinLayers>::value, "type of JoinLayerType must equal to JoinLayers");
```